### ISSUE TITLE
[RFC] Enable strict errors.

### ISSUE THREAD

**Issue by dantleech** [bolt/bolt#5504]
By default bolt is very tolerant of errors. Strict variables is turned off in Twig for example.

Every error that goes unnoticed increases technical debt - either you have avoidable bugs (e.g. typos in var names) or you turn on strict error reporting later only to find that many pages contain errors.

This example from Bolt Forms (due to a configuration error on my part):

```
Neither the property "submit" nor one of the methods "submit()", "getsubmit()"/"issubmit()" or "__call()" exist and have public access in class "Symfony\Component\Form\FormView" in "boltforms_form.twig" at line 72
```

And lots of examples from my own project when I turned on strict variables.

If variables are undefined, then this is almost certainly a programming error, and therefore the application is potentially not behaving as intended. When in development mode applications should shout as loudly and verbosely as the can whenever an error is encountered.

I would suggest that by default bolt has strict_variables enabled when in debug mode.

**Comment by GawainLynch** [comment]
Well for me it comes down to separation of concerns. There are times/sites that I want debug on and `strict_variables` off … and the setting for `debug:` and `strict_variables:` are [all close together](https://github.com/bolt/bolt/blob/release/3.0/app/config/config.yml.dist#L252-L274) in a "debug" grouping of settings.

**Comment by GawainLynch** [comment]
Also for the core team (not sure is/where documented) we ask them to put both on when doing dev work as well as:

```yaml
developer_notices: true
debug_error_level -1
```
Might need to get that sorted doc-wise.

**Comment by GawainLynch** [comment]
Just occurred to me that this is also a good habit to add to a dev deployer:

```
$ ./app/nut config:get strict_variables
strict_variables: 
$ ./app/nut config:set strict_variables true
New value for strict_variables: true was successful. File updated.
$ ./app/nut config:get strict_variables
strict_variables: 1
```

**Comment by SahAssar** [comment]
:+1: for strict variables and developer notices in debug mode, but if so I'd say that debug mode should not be enabled by default.

**Comment by GawainLynch** [comment]
It is also a balancing act. We want a new Bolt install user to be able to experience the debug bar, profiler, etc to sell off what Bolt can do. If they've just then grabbed a theme off of the market place — that works fine in non-strict mode (a.k.a. less annoying to the average user mode) — and ***boom*** "Bolt is broken on a fresh install… This sucks".

Increasingly thinking this is a documentation issue.

**Comment by dantleech** [comment]
Yeah, I can see the "first impression" issue, but think that if a theme has an error, then that is the theme's problem - developers should not release themes which depend on unavailable variables, in the same way they shouldn't release themes with parsing errors. Enabling strict variables by default (in debug mode) would help prevent this. (maybe in the long run some basic CI for themes could also help).

People will enable this eventually (I hope) and when they do they will be confronted with a whole bundle of errors which they didn't know they had. Hiding issues has short term benefits, but in the long run it doesn't pay off.

Just my 2c.

**Comment by bobdenotter** [comment]
> `developer_notices: true`
> `debug_error_level: -1`

And don't forget also: 

    strict_variables: true

**Comment by GawainLynch** [comment]
On re-reflection, I am thinking the best approach here might be a (set of?) page(s) with advice on setting up your environment for:

  * Theme development
  * Extension development
  * Core development

**Comment by SahAssar** [comment]
Well, it's not just themes in the store... For most usual designers, theme developers and bolt users doing something like `{% if array.key %}` seems pretty normal, while that wouldn't work if that key didn't exist (with strict_variables). To me this definitely seems like either we should have stricter debug rules and not have debug on by default or that we need better docs.

**Comment by GawainLynch** [comment]
On this one, better docs … if sites start crashing when people are "kicking tires" it will look like it is Bolt's fault.

However, a developer should easily find out this setting is there and should be used when starting out.

**Comment by dantleech** [comment]
Well, if you enable strict by default then people will not be doing `array.key` as a test, they should be doing `array.key is defined` and nobody should be kicking tires.

My fear here is that by the time a developer finds this option they have already coded much of the site, and much of it actually has errors (as was the case with me).

A per-environment configuration would help here, as with Symfony. i.e. in `dev` this is enabled, but on `prod` it is off.

**Comment by SahAssar** [comment]
Strict variables are now enabled by default in 3.2 (if on a composer install, in a zip/tar install it is off), the exception handling has been improved, so many of these will be surfaced now.

Some of the exception handling was rolled back because of too much drama (sorry you-know-who), but IIRC it will be reintroduced in the future. Basically I'm saying that much of this is fixed (and the rest will be in 4.0 IIRC), and I'm closing :)
